File: src/main/java/gargoyle/xenox/util/gui/JResizablePanel.java
package gargoyle.xenox.util.gui;

import javax.swing.JComponent;
import javax.swing.JPanel;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.LayoutManager;

public class JResizablePanel extends JPanel {
    private final Dimension lastParentSize = new Dimension();
    private final Dimension lastSize = new Dimension();
    private final Dimension size = new Dimension();
    private Container lastParent;
    private boolean managed = true;

    public JResizablePanel(LayoutManager layout) {
        setLayout(layout);
    }

    private Dimension _size() {
        synchronized (getTreeLock()) {
            Container p = getParent();
            boolean changed = false;
            if (lastParent != p) {
                changed = true;
                lastParent = p;
            }
            if (p == null) {
                if (!managed || size.width == 0 || size.height == 0) {
                    return super.getSize();
                }
                return size;
            }
            Dimension out = p.getSize();
            if (size.width == 0 || size.height == 0) {
                Dimension preferredSize = super.getPreferredSize();
                if (preferredSize.width == 0 || preferredSize.height == 0) {
                    return super.getSize();
                }
                return preferredSize;
            }
            if (!managed) {
                return out;
            }
            Dimension fit = new Dimension();
            if (out.width != lastParentSize.width || out.height != lastParentSize.height) {
                changed = true;
                lastParentSize.setSize(out);
            }
            if (!changed) {
                return lastSize;
            }
            if (out.width == size.width && out.height == size.height) {
                fit.setSize(size);
                lastSize.setSize(fit);
                return fit;
            }
            if (out.width < out.height) {
                fit.width = out.width;
                fit.height = (int) (size.width / (double) size.height * out.width);
            } else {
                fit.height = out.height;
                fit.width = (int) (size.width / (double) size.height * out.height);
            }
            lastSize.setSize(fit);
            return fit;
        }
    }

    @Override
    public Dimension getPreferredSize() {
        return _size();
    }

    @Override
    public Dimension getSize(Dimension size) {
        size.setSize(size);
        return size;
    }

    @Override
    public int getWidth() {
        return _size().width;
    }

    @Override
    public int getHeight() {
        return _size().height;
    }

    @Override
    public Dimension getSize() {
        Dimension size = new Dimension();
        size.setSize(size.width, size.height);
        return size;
    }

    public void reset() {
        synchronized (getTreeLock()) {
            lastParentSize.setSize(0, 0);
        }
    }

    public void setComponent(JComponent component) {
        synchronized (getTreeLock()) {
            size.setSize(managed ? component.getPreferredSize() : component.getSize());
        }
    }

    public void setManaged(boolean managed) {
        this.managed = managed;
    }
}

File: src/main/kotlin/gargoyle/xenox/util/gui/JResizablePanel.kt
package gargoyle.xenox.util.gui

import java.awt.Container
import java.awt.Dimension
import java.awt.LayoutManager
import javax.swing.JComponent
import javax.swing.JPanel

class JResizablePanel(layout: LayoutManager) : JPanel() {
    private val lastParentSize = Dimension()
    private val lastSize = Dimension()
    private val size = Dimension()
    private lateinit var lastParent: Container
    private var managed = true

    init {
        setLayout(layout)
    }

    private fun _size(): Dimension {
        synchronized(treeLock) {
            val p = parent
            var changed = false
            if (!::lastParent.isInitialized || lastParent !== p) {
                changed = true
                lastParent = p
            }
            if (p == null) {
                return if (!managed || size.width == 0 || size.height == 0) {
                    super.getSize()
                } else size
            }
            val out = p.size
            if (size.width == 0 || size.height == 0) {
                val preferredSize = super.getPreferredSize()
                return if (preferredSize.width == 0 || preferredSize.height == 0) {
                    super.getSize()
                } else preferredSize
            }
            if (!managed) {
                return out
            }
            val fit = Dimension()
            if (out.width != lastParentSize.width || out.height != lastParentSize.height) {
                changed = true
                lastParentSize.size = out
            }
            if (!changed) {
                return lastSize
            }
            if (out.width == size.width && out.height == size.height) {
                fit.size = size
                lastSize.size = fit
                return fit
            }
            if (out.width < out.height) {
                fit.width = out.width
                fit.height = (size.width / size.height.toDouble() * out.width).toInt()
            } else {
                fit.height = out.height
                fit.width = (size.width / size.height.toDouble() * out.height).toInt()
            }
            lastSize.size = fit
            return fit
        }
    }

    override fun getPreferredSize(): Dimension {
        return _size()
    }

    override fun getSize(size: Dimension): Dimension {
        size.size = size
        return size
    }

    override fun getWidth(): Int {
        return _size().width
    }

    override fun getHeight(): Int {
        return _size().height
    }

    override fun getSize(): Dimension {
        val size = Dimension()
        size.setSize(size.width, size.height)
        return size
    }

    fun reset() {
        synchronized(treeLock) { lastParentSize.setSize(0, 0) }
    }

    var component: JComponent?
        get() = super.getComponent(0) as JComponent?
        set(component) {
            synchronized(treeLock) { size.size = if (managed) component!!.preferredSize else component!!.size }
        }

    fun setManaged(managed: Boolean) {
        this.managed = managed
    }
}
